File: b4ae-android-app/app/src/main/java/com/b4ae/B4AEEnhanced.kt
package com.b4ae

/**
 * Enhanced B4AE Android SDK - Full quantum-safe cryptography support
 * 
 * Provides comprehensive quantum-resistant secure communication for Android applications.
 * 
 * Features:
 * - Post-quantum cryptography (Kyber-1024, Dilithium5)
 * - Hybrid cryptography (X25519, Ed25519)
 * - AES-256-GCM encryption
 * - Complete handshake protocol
 * - Session management
 * - Metadata protection
 */
object B4AE {
    
    // Constants
    const val KEY_SIZE = 32
    const val NONCE_SIZE = 12
    const val KYBER_PUBLIC_KEY_SIZE = 1568
    const val KYBER_SECRET_KEY_SIZE = 3168
    const val KYBER_CIPHERTEXT_SIZE = 1568
    const val KYBER_SHARED_SECRET_SIZE = 32
    const val DILITHIUM_PUBLIC_KEY_SIZE = 2592
    const val DILITHIUM_SECRET_KEY_SIZE = 4864
    const val DILITHIUM_SIGNATURE_SIZE = 4595
    
    // Security profiles
    enum class SecurityProfile(val value: Int) {
        STANDARD(0),
        HIGH(1),
        MAXIMUM(2),
        ENTERPRISE(3)
    }
    
    // Exception classes
    class B4AEException(message: String) : Exception(message)
    class B4AEHandshakeException(message: String) : Exception(message)
    class B4AESessionException(message: String) : Exception(message)
    
    // Data classes
    data class Keypair(
        val publicKey: ByteArray,
        val secretKey: ByteArray
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as Keypair
            return publicKey.contentEquals(other.publicKey) && secretKey.contentEquals(other.secretKey)
        }
        
        override fun hashCode(): Int {
            var result = publicKey.contentHashCode()
            result = 31 * result + secretKey.contentHashCode()
            return result
        }
    }
    
    data class EncapsulationResult(
        val ciphertext: ByteArray,
        val sharedSecret: ByteArray
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as EncapsulationResult
            return ciphertext.contentEquals(other.ciphertext) && sharedSecret.contentEquals(other.sharedSecret)
        }
        
        override fun hashCode(): Int {
            var result = ciphertext.contentHashCode()
            result = 31 * result + sharedSecret.contentHashCode()
            return result
        }
    }
    
    data class EncryptedMessage(
        val nonce: ByteArray,
        val ciphertext: ByteArray
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as EncryptedMessage
            return nonce.contentEquals(other.nonce) && ciphertext.contentEquals(other.ciphertext)
        }
        
        override fun hashCode(): Int {
            var result = nonce.contentHashCode()
            result = 31 * result + ciphertext.contentHashCode()
            return result
        }
        
        fun toByteArray(): ByteArray {
            return nonce + ciphertext
        }
        
        companion object {
            fun fromByteArray(data: ByteArray): EncryptedMessage {
                if (data.size < NONCE_SIZE) {
                    throw B4AEException("Invalid encrypted message format")
                }
                val nonce = data.copyOfRange(0, NONCE_SIZE)
                val ciphertext = data.copyOfRange(NONCE_SIZE, data.size)
                return EncryptedMessage(nonce, ciphertext)
            }
        }
    }
    
    data class SessionInfo(
        val peerId: String,
        val isEstablished: Boolean,
        val messagesExchanged: Long,
        val lastActivity: Long
    )
    
    // Native library loading
    init {
        try {
            System.loadLibrary("b4ae_android")
        } catch (e: UnsatisfiedLinkError) {
            throw B4AEException("Failed to load B4AE native library: ${e.message}")
        }
    }
    
    // Native method declarations
    private external fun nativeInit(securityProfile: Int): Long
    private external fun nativeCleanup(clientPtr: Long)
    private external fun nativeGenerateKeypair(clientPtr: Long): ByteArray?
    private external fun nativeEncapsulate(clientPtr: Long, publicKey: ByteArray): ByteArray?
    private external fun nativeDecapsulate(clientPtr: Long, ciphertext: ByteArray, secretKey: ByteArray): ByteArray?
    private external fun nativeSign(clientPtr: Long, data: ByteArray, secretKey: ByteArray): ByteArray?
    private external fun nativeVerify(clientPtr: Long, signature: ByteArray, data: ByteArray, publicKey: ByteArray): Boolean
    private external fun nativeEncrypt(clientPtr: Long, key: ByteArray, plaintext: ByteArray): ByteArray?
    private external fun nativeDecrypt(clientPtr: Long, key: ByteArray, encrypted: ByteArray): ByteArray?
    private external fun nativeHandshake(clientPtr: Long, peerId: ByteArray): ByteArray?
    private external fun nativeCompleteHandshake(clientPtr: Long, peerId: ByteArray, handshakeData: ByteArray): Boolean
    private external fun nativeEncryptMessage(clientPtr: Long, peerId: ByteArray, message: ByteArray): ByteArray?
    private external fun nativeDecryptMessage(clientPtr: Long, peerId: ByteArray, encryptedMessages: ByteArray): ByteArray?
    private external fun nativeGetVersion(): String
    private external fun nativeGetSecurityInfo(clientPtr: Long): String
    
    // Public API methods
    
    /**
     * Initialize B4AE client with security profile
     */
    fun initialize(profile: SecurityProfile = SecurityProfile.STANDARD): B4AEClient {
        val clientPtr = nativeInit(profile.value)
        if (clientPtr == 0L) {
            throw B4AEException("Failed to initialize B4AE client")
        }
        return B4AEClient(clientPtr, profile)
    }
    
    /**
     * Get B4AE library version
     */
    fun getVersion(): String = nativeGetVersion()
    
    /**
     * Get security information
     */
    fun getSecurityInfo(client: B4AEClient): String = nativeGetSecurityInfo(client.clientPtr)
    
    /**
     * Generate quantum-safe keypair (Kyber-1024)
     */
    fun generateKeypair(client: B4AEClient): Keypair {
        val keypairBytes = nativeGenerateKeypair(client.clientPtr)
            ?: throw B4AEException("Failed to generate keypair")
        
        if (keypairBytes.size != KYBER_PUBLIC_KEY_SIZE + KYBER_SECRET_KEY_SIZE) {
            throw B4AEException("Invalid keypair size: ${keypairBytes.size}")
        }
        
        val publicKey = keypairBytes.copyOfRange(0, KYBER_PUBLIC_KEY_SIZE)
        val secretKey = keypairBytes.copyOfRange(KYBER_PUBLIC_KEY_SIZE, keypairBytes.size)
        
        return Keypair(publicKey, secretKey)
    }
    
    /**
     * Perform key encapsulation (Kyber-1024)
     */
    fun encapsulate(client: B4AEClient, publicKey: ByteArray): EncapsulationResult {
        if (publicKey.size != KYBER_PUBLIC_KEY_SIZE) {
            throw B4AEException("Invalid public key size: ${publicKey.size}, expected: $KYBER_PUBLIC_KEY_SIZE")
        }
        
        val resultBytes = nativeEncapsulate(client.clientPtr, publicKey)
            ?: throw B4AEException("Failed to encapsulate")
        
        if (resultBytes.size != KYBER_CIPHERTEXT_SIZE + KYBER_SHARED_SECRET_SIZE) {
            throw B4AEException("Invalid encapsulation result size: ${resultBytes.size}")
        }
        
        val ciphertext = resultBytes.copyOfRange(0, KYBER_CIPHERTEXT_SIZE)
        val sharedSecret = resultBytes.copyOfRange(KYBER_CIPHERTEXT_SIZE, resultBytes.size)
        
        return EncapsulationResult(ciphertext, sharedSecret)
    }
    
    /**
     * Decapsulate shared secret (Kyber-1024)
     */
    fun decapsulate(client: B4AEClient, ciphertext: ByteArray, secretKey: ByteArray): ByteArray {
        if (ciphertext.size != KYBER_CIPHERTEXT_SIZE) {
            throw B4AEException("Invalid ciphertext size: ${ciphertext.size}, expected: $KYBER_CIPHERTEXT_SIZE")
        }
        if (secretKey.size != KYBER_SECRET_KEY_SIZE) {
            throw B4AEException("Invalid secret key size: ${secretKey.size}, expected: $KYBER_SECRET_KEY_SIZE")
        }
        
        return nativeDecapsulate(client.clientPtr, ciphertext, secretKey)
            ?: throw B4AEException("Failed to decapsulate")
    }
    
    /**
     * Sign data with Dilithium5 (post-quantum signature)
     */
    fun sign(client: B4AEClient, data: ByteArray, secretKey: ByteArray): ByteArray {
        if (secretKey.size != DILITHIUM_SECRET_KEY_SIZE) {
            throw B4AEException("Invalid secret key size: ${secretKey.size}, expected: $DILITHIUM_SECRET_KEY_SIZE")
        }
        
        return nativeSign(client.clientPtr, data, secretKey)
            ?: throw B4AEException("Failed to sign data")
    }
    
    /**
     * Verify Dilithium5 signature
     */
    fun verify(client: B4AEClient, signature: ByteArray, data: ByteArray, publicKey: ByteArray): Boolean {
        if (publicKey.size != DILITHIUM_PUBLIC_KEY_SIZE) {
            throw B4AEException("Invalid public key size: ${publicKey.size}, expected: $DILITHIUM_PUBLIC_KEY_SIZE")
        }
        if (signature.size != DILITHIUM_SIGNATURE_SIZE) {
            throw B4AEException("Invalid signature size: ${signature.size}, expected: $DILITHIUM_SIGNATURE_SIZE")
        }
        
        return nativeVerify(client.clientPtr, signature, data, publicKey)
    }
    
    /**
     * Encrypt data with AES-256-GCM
     */
    fun encrypt(client: B4AEClient, key: ByteArray, plaintext: ByteArray): EncryptedMessage {
        if (key.size != KEY_SIZE) {
            throw B4AEException("Invalid key size: ${key.size}, expected: $KEY_SIZE")
        }
        
        val encryptedBytes = nativeEncrypt(client.clientPtr, key, plaintext)
            ?: throw B4AEException("Failed to encrypt data")
        
        return EncryptedMessage.fromByteArray(encryptedBytes)
    }
    
    /**
     * Decrypt data with AES-256-GCM
     */
    fun decrypt(client: B4AEClient, key: ByteArray, encryptedMessage: EncryptedMessage): ByteArray {
        if (key.size != KEY_SIZE) {
            throw B4AEException("Invalid key size: ${key.size}, expected: $KEY_SIZE")
        }
        
        val encryptedBytes = encryptedMessage.toByteArray()
        return nativeDecrypt(client.clientPtr, key, encryptedBytes)
            ?: throw B4AEException("Failed to decrypt data")
    }
    
    /**
     * Perform quantum-safe handshake
     */
    fun performHandshake(client: B4AEClient, peerId: String): ByteArray {
        val peerIdBytes = peerId.toByteArray(Charsets.UTF_8)
        
        return nativeHandshake(client.clientPtr, peerIdBytes)
            ?: throw B4AEHandshakeException("Failed to initiate handshake")
    }
    
    /**
     * Complete handshake response
     */
    fun completeHandshake(client: B4AEClient, peerId: String, handshakeData: ByteArray): Boolean {
        val peerIdBytes = peerId.toByteArray(Charsets.UTF_8)
        
        return nativeCompleteHandshake(client.clientPtr, peerIdBytes, handshakeData)
    }
    
    /**
     * Encrypt message with established session
     */
    fun encryptMessage(client: B4AEClient, peerId: String, message: ByteArray): ByteArray {
        val peerIdBytes = peerId.toByteArray(Charsets.UTF_8)
        
        return nativeEncryptMessage(client.clientPtr, peerIdBytes, message)
            ?: throw B4AESessionException("Failed to encrypt message: session not established or invalid")
    }
    
    /**
     * Decrypt message with established session
     */
    fun decryptMessage(client: B4AEClient, peerId: String, encryptedMessages: ByteArray): ByteArray {
        val peerIdBytes = peerId.toByteArray(Charsets.UTF_8)
        
        return nativeDecryptMessage(client.clientPtr, peerIdBytes, encryptedMessages)
            ?: throw B4AESessionException("Failed to decrypt message: session not established or invalid")
    }
    
    /**
     * B4AE Client wrapper
     */
    class B4AEClient(
        internal val clientPtr: Long,
        val securityProfile: SecurityProfile
    ) {
        private var isDisposed = false
        
        /**
         * Check if client is valid
         */
        fun isValid(): Boolean = !isDisposed && clientPtr != 0L
        
        /**
         * Dispose client and free resources
         */
        fun dispose() {
            if (!isDisposed && clientPtr != 0L) {
                nativeCleanup(clientPtr)
                isDisposed = true
            }
        }
        
        /**
         * Auto-dispose on garbage collection
         */
        protected fun finalize() {
            dispose()
        }
    }
    
    /**
     * Utility functions
     */
    object Utils {
        /**
         * Generate cryptographically secure random bytes
         */
        fun generateRandomBytes(size: Int): ByteArray {
            val bytes = ByteArray(size)
            java.security.SecureRandom().nextBytes(bytes)
            return bytes
        }
        
        /**
         * Convert string to UTF-8 bytes
         */
        fun stringToBytes(text: String): ByteArray = text.toByteArray(Charsets.UTF_8)
        
        /**
         * Convert bytes to UTF-8 string
         */
        fun bytesToString(bytes: ByteArray): String = String(bytes, Charsets.UTF_8)
        
        /**
         * Check if two byte arrays are equal (constant-time)
         */
        fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
            if (a.size != b.size) return false
            var result = 0
            for (i in a.indices) {
                result = result or (a[i].toInt() xor b[i].toInt())
            }
            return result == 0
        }
    }
}

File: bindings/swift/Sources/B4AE/B4AEEnhanced.swift
//
//  B4AEEnhanced.swift
//  Enhanced B4AE iOS SDK - Full quantum-safe cryptography support
//

import Foundation

/// B4AE Security Profiles
public enum B4AESecurityProfile: Int32 {
    case standard = 0
    case high = 1
    case maximum = 2
    case enterprise = 3
}

/// B4AE Keypair structure
public struct B4AEKeypair {
    public let publicKey: Data
    public let secretKey: Data
    
    public init(publicKey: Data, secretKey: Data) {
        self.publicKey = publicKey
        self.secretKey = secretKey
    }
}

/// B4AE Encapsulation Result
public struct B4AEEncapsulationResult {
    public let ciphertext: Data
    public let sharedSecret: Data
    
    public init(ciphertext: Data, sharedSecret: Data) {
        self.ciphertext = ciphertext
        self.sharedSecret = sharedSecret
    }
}

/// B4AE Encrypted Message
public struct B4AEEncryptedMessage {
    public let nonce: Data
    public let ciphertext: Data
    
    public init(nonce: Data, ciphertext: Data) {
        self.nonce = nonce
        self.ciphertext = ciphertext
    }
    
    public func toData() -> Data {
        return nonce + ciphertext
    }
    
    public static func fromData(_ data: Data) throws -> B4AEEncryptedMessage {
        guard data.count >= B4AEConstants.nonceSize else {
            throw B4AEError.invalidFormat
        }
        
        let nonce = data.prefix(B4AEConstants.nonceSize)
        let ciphertext = data.suffix(from: B4AEConstants.nonceSize)
        
        return B4AEEncryptedMessage(nonce: nonce, ciphertext: ciphertext)
    }
}

/// B4AE Constants
public struct B4AEConstants {
    public static let keySize = 32
    public static let nonceSize = 12
    public static let kyberPublicKeySize = 1568
    public static let kyberSecretKeySize = 3168
    public static let kyberCiphertextSize = 1568
    public static let kyberSharedSecretSize = 32
    public static let dilithiumPublicKeySize = 2592
    public static let dilithiumSecretKeySize = 4864
    public static let dilithiumSignatureSize = 4595
}

/// B4AE Errors
public enum B4AEError: Error {
    case initializationFailed
    case keyGenerationFailed
    case encryptionFailed
    case decryptionFailed
    case handshakeFailed
    case sessionNotEstablished
    case invalidFormat
    case invalidKeySize
    case invalidParameter
    case nativeError(String)
}

/// Enhanced B4AE iOS SDK
public class B4AEEnhanced {
    
    // MARK: - Properties
    
    private var clientPtr: OpaquePointer?
    private let securityProfile: B4AESecurityProfile
    private let queue = DispatchQueue(label: "com.b4ae.enhanced", qos: .userInitiated)
    
    // MARK: - Initialization
    
    private init(clientPtr: OpaquePointer?, securityProfile: B4AESecurityProfile) {
        self.clientPtr = clientPtr
        self.securityProfile = securityProfile
    }
    
    /// Initialize B4AE with security profile
    public static func initialize(profile: B4AESecurityProfile = .standard) throws -> B4AEEnhanced {
        let clientPtr = b4ae_init_enhanced(profile.rawValue)
        
        guard clientPtr != nil else {
            throw B4AEError.initializationFailed
        }
        
        return B4AEEnhanced(clientPtr: clientPtr, securityProfile: profile)
    }
    
    /// Cleanup resources
    deinit {
        if let clientPtr = clientPtr {
            b4ae_cleanup_enhanced(clientPtr)
        }
    }
    
    // MARK: - Public API
    
    /// Get library version
    public static func getVersion() -> String {
        guard let version = b4ae_get_version_enhanced() else {
            return "unknown"
        }
        return String(cString: version)
    }
    
    /// Get security information
    public func getSecurityInfo() -> String {
        guard let clientPtr = clientPtr else {
            return "Client not initialized"
        }
        
        guard let info = b4ae_get_security_info_enhanced(clientPtr) else {
            return "Security info unavailable"
        }
        
        return String(cString: info)
    }
    
    /// Generate quantum-safe keypair (Kyber-1024)
    public func generateKeypair() throws -> B4AEKeypair {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        var publicKeyData = Data()
        var secretKeyData = Data()
        
        let result = publicKeyData.withUnsafeMutableBytes { publicKeyPtr in
            secretKeyData.withUnsafeMutableBytes { secretKeyPtr in
                b4ae_generate_keypair_enhanced(
                    clientPtr,
                    publicKeyPtr.bindMemory(to: UInt8.self).baseAddress,
                    secretKeyPtr.bindMemory(to: UInt8.self).baseAddress
                )
            }
        }
        
        guard result == 0 else {
            throw B4AEError.keyGenerationFailed
        }
        
        return B4AEKeypair(publicKey: publicKeyData, secretKey: secretKeyData)
    }
    
    /// Perform key encapsulation (Kyber-1024)
    public func encapsulate(publicKey: Data) throws -> B4AEEncapsulationResult {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard publicKey.count == B4AEConstants.kyberPublicKeySize else {
            throw B4AEError.invalidKeySize
        }
        
        var ciphertextData = Data(count: B4AEConstants.kyberCiphertextSize)
        var sharedSecretData = Data(count: B4AEConstants.kyberSharedSecretSize)
        
        let result = publicKey.withUnsafeBytes { publicKeyPtr in
            ciphertextData.withUnsafeMutableBytes { ciphertextPtr in
                sharedSecretData.withUnsafeMutableBytes { sharedSecretPtr in
                    b4ae_encapsulate_enhanced(
                        clientPtr,
                        publicKeyPtr.bindMemory(to: UInt8.self).baseAddress,
                        ciphertextPtr.bindMemory(to: UInt8.self).baseAddress,
                        sharedSecretPtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        guard result == 0 else {
            throw B4AEError.encryptionFailed
        }
        
        return B4AEEncapsulationResult(ciphertext: ciphertextData, sharedSecret: sharedSecretData)
    }
    
    /// Decapsulate shared secret (Kyber-1024)
    public func decapsulate(ciphertext: Data, secretKey: Data) throws -> Data {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard ciphertext.count == B4AEConstants.kyberCiphertextSize else {
            throw B4AEError.invalidKeySize
        }
        
        guard secretKey.count == B4AEConstants.kyberSecretKeySize else {
            throw B4AEError.invalidKeySize
        }
        
        var sharedSecretData = Data(count: B4AEConstants.kyberSharedSecretSize)
        
        let result = ciphertext.withUnsafeBytes { ciphertextPtr in
            secretKey.withUnsafeBytes { secretKeyPtr in
                sharedSecretData.withUnsafeMutableBytes { sharedSecretPtr in
                    b4ae_decapsulate_enhanced(
                        clientPtr,
                        ciphertextPtr.bindMemory(to: UInt8.self).baseAddress,
                        secretKeyPtr.bindMemory(to: UInt8.self).baseAddress,
                        sharedSecretPtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        guard result == 0 else {
            throw B4AEError.decryptionFailed
        }
        
        return sharedSecretData
    }
    
    /// Sign data with Dilithium5 (post-quantum signature)
    public func sign(data: Data, secretKey: Data) throws -> Data {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard secretKey.count == B4AEConstants.dilithiumSecretKeySize else {
            throw B4AEError.invalidKeySize
        }
        
        var signatureData = Data(count: B4AEConstants.dilithiumSignatureSize)
        
        let result = data.withUnsafeBytes { dataPtr in
            secretKey.withUnsafeBytes { secretKeyPtr in
                signatureData.withUnsafeMutableBytes { signaturePtr in
                    b4ae_sign_enhanced(
                        clientPtr,
                        dataPtr.bindMemory(to: UInt8.self).baseAddress,
                        UInt(data.count),
                        secretKeyPtr.bindMemory(to: UInt8.self).baseAddress,
                        signaturePtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        guard result == 0 else {
            throw B4AEError.encryptionFailed
        }
        
        return signatureData
    }
    
    /// Verify Dilithium5 signature
    public func verify(signature: Data, data: Data, publicKey: Data) throws -> Bool {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard signature.count == B4AEConstants.dilithiumSignatureSize else {
            throw B4AEError.invalidKeySize
        }
        
        guard publicKey.count == B4AEConstants.dilithiumPublicKeySize else {
            throw B4AEError.invalidKeySize
        }
        
        let result = signature.withUnsafeBytes { signaturePtr in
            data.withUnsafeBytes { dataPtr in
                publicKey.withUnsafeBytes { publicKeyPtr in
                    b4ae_verify_enhanced(
                        clientPtr,
                        signaturePtr.bindMemory(to: UInt8.self).baseAddress,
                        dataPtr.bindMemory(to: UInt8.self).baseAddress,
                        UInt(data.count),
                        publicKeyPtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        return result != 0
    }
    
    /// Encrypt with AES-256-GCM
    public func encrypt(key: Data, plaintext: Data) throws -> B4AEEncryptedMessage {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard key.count == B4AEConstants.keySize else {
            throw B4AEError.invalidKeySize
        }
        
        var encryptedData = Data(count: B4AEConstants.nonceSize + plaintext.count + 16) // nonce + ciphertext + tag
        
        let result = key.withUnsafeBytes { keyPtr in
            plaintext.withUnsafeBytes { plaintextPtr in
                encryptedData.withUnsafeMutableBytes { encryptedPtr in
                    b4ae_encrypt_enhanced(
                        clientPtr,
                        keyPtr.bindMemory(to: UInt8.self).baseAddress,
                        plaintextPtr.bindMemory(to: UInt8.self).baseAddress,
                        UInt(plaintext.count),
                        encryptedPtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        guard result == 0 else {
            throw B4AEError.encryptionFailed
        }
        
        return try B4AEEncryptedMessage.fromData(encryptedData)
    }
    
    /// Decrypt with AES-256-GCM
    public func decrypt(key: Data, encryptedMessage: B4AEEncryptedMessage) throws -> Data {
        guard let clientPtr = clientPtr else {
            throw B4AEError.initializationFailed
        }
        
        guard key.count == B4AEConstants.keySize else {
            throw B4AEError.invalidKeySize
        }
        
        let encryptedData = encryptedMessage.toData()
        guard encryptedData.count > B4AEConstants.nonceSize else {
            throw B4AEError.invalidFormat
        }
        
        var decryptedData = Data(count: encryptedData.count - B4AEConstants.nonceSize - 16) // Remove nonce and tag
        
        let result = key.withUnsafeBytes { keyPtr in
            encryptedData.withUnsafeBytes { encryptedPtr in
                decryptedData.withUnsafeMutableBytes { decryptedPtr in
                    b4ae_decrypt_enhanced(
                        clientPtr,
                        keyPtr.bindMemory(to: UInt8.self).baseAddress,
                        encryptedPtr.bindMemory(to: UInt8.self).baseAddress,
                        UInt(encryptedData.count),
                        decryptedPtr.bindMemory(to: UInt8.self).baseAddress
                    )
                }
            }
        }
        
        guard result == 0 else {
            throw B4AEError.decryptionFailed
        }
        
        return decryptedData
    }
    
    // MARK: - Utility Methods
    
    /// Generate cryptographically secure random bytes
    public static func generateRandomBytes(count: Int) -> Data {
        var bytes = Data(count: count)
        bytes.withUnsafeMutableBytes { bytesPtr in
            SecRandomCopyBytes(kSecRandomDefault, count, bytesPtr.bindMemory(to: UInt8.self).baseAddress!)
        }
        return bytes
    }
    
    /// Convert string to UTF-8 data
    public static func stringToData(_ string: String) -> Data {
        return string.data(using: .utf8) ?? Data()
    }
    
    /// Convert data to UTF-8 string
    public static func dataToString(_ data: Data) -> String {
        return String(data: data, encoding: .utf8) ?? ""
    }
    
    /// Constant-time comparison of data
    public static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        
        var result = 0
        for i in 0..<a.count {
            result |= Int(a[i]) ^ Int(b[i])
        }
        
        return result == 0
    }
}

// MARK: - Native Function Declarations

// Native function declarations would be implemented in the bridging header
// These are placeholder declarations that would link to the actual C implementation

private func b4ae_init_enhanced(_ profile: Int32) -> OpaquePointer? {
    // Implementation would link to native C/Rust code
    return nil
}

private func b4ae_cleanup_enhanced(_ client: OpaquePointer) {
    // Implementation would link to native C/Rust code
}

private func b4ae_get_version_enhanced() -> UnsafePointer<CChar>? {
    // Implementation would link to native C/Rust code
    return nil
}

private func b4ae_get_security_info_enhanced(_ client: OpaquePointer) -> UnsafePointer<CChar>? {
    // Implementation would link to native C/Rust code
    return nil
}

private func b4ae_generate_keypair_enhanced(
    _ client: OpaquePointer,
    _ publicKey: UnsafeMutablePointer<UInt8>,
    _ secretKey: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_encapsulate_enhanced(
    _ client: OpaquePointer,
    _ publicKey: UnsafePointer<UInt8>,
    _ ciphertext: UnsafeMutablePointer<UInt8>,
    _ sharedSecret: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_decapsulate_enhanced(
    _ client: OpaquePointer,
    _ ciphertext: UnsafePointer<UInt8>,
    _ secretKey: UnsafePointer<UInt8>,
    _ sharedSecret: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_sign_enhanced(
    _ client: OpaquePointer,
    _ data: UnsafePointer<UInt8>,
    _ dataLen: UInt,
    _ secretKey: UnsafePointer<UInt8>,
    _ signature: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_verify_enhanced(
    _ client: OpaquePointer,
    _ signature: UnsafePointer<UInt8>,
    _ data: UnsafePointer<UInt8>,
    _ dataLen: UInt,
    _ publicKey: UnsafePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_encrypt_enhanced(
    _ client: OpaquePointer,
    _ key: UnsafePointer<UInt8>,
    _ plaintext: UnsafePointer<UInt8>,
    _ plaintextLen: UInt,
    _ encrypted: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}

private func b4ae_decrypt_enhanced(
    _ client: OpaquePointer,
    _ key: UnsafePointer<UInt8>,
    _ encrypted: UnsafePointer<UInt8>,
    _ encryptedLen: UInt,
    _ decrypted: UnsafeMutablePointer<UInt8>
) -> Int32 {
    // Implementation would link to native C/Rust code
    return -1
}